---
title: 'GLORI EDA'
output:
  html_document: default
---

GLORI EDA
========================================================

```{r echo=FALSE, message=FALSE, warning=FALSE, packages}
# Load all of the packages that you end up using
# in your analysis in this code chunk.
library(tidyverse)
library(readxl)
library(biomaRt)
library(fuzzyjoin)
library(ggpubr) 
library(ggrepel)
library(latex2exp)
# GFF/GTF related libs
library(rtracklayer) # Interacting with UCSC genome browser
library(GenomicRanges)
library(plyranges)
library(GenomicFeatures)
library(ensembldb) 
# For dealing with bamfiles
library(Rsamtools)
```

```{r echo=FALSE, Load_the_Data}
# Load the Data
## Can be changed
data_folder = '/Users/faqih/Documents/UCPH/Thesis/code/data/GLORI_data/'

overlap_repl_sites <- readxl::read_excel(paste(data_folder, 'identified_m6A_HEK29T.xlsx', sep=''))
overlap_repl_sites$Sites <- as.integer(overlap_repl_sites$Sites)
# Int conversion
overlap_repl_sites$AGCov_rep1 <- as.integer(overlap_repl_sites$AGCov_rep1)
overlap_repl_sites$AGCov_rep2 <- as.integer(overlap_repl_sites$AGCov_rep2)
overlap_repl_sites <- overlap_repl_sites |> 
  rename_with(tolower)
# Average over replicates
overlap_repl_sites <- overlap_repl_sites |> 
  mutate(avg_m6a = (m6a_level_rep1 + m6a_level_rep1)/2)


## Heat Shock MEF Data 
heat_shock_MEF_up <- readxl::read_excel(paste(data_folder, 
                                                       'heat_shock_m6A_sites_MEF.xlsx', 
                                                       sep=''),
                                                 sheet = 'HS UP m6A')
heat_shock_MEF_down <- readxl::read_excel(paste(data_folder, 
                                                       'heat_shock_m6A_sites_MEF.xlsx', 
                                                       sep=''),
                                                 sheet = 'HS Down m6A')
heat_shock_MEF_unchange <- readxl::read_excel(paste(data_folder, 
                                                       'heat_shock_m6A_sites_MEF.xlsx', 
                                                       sep=''),
                                                 sheet = 'HS un-change m6A')

heat_shock_MEF <- rbind(heat_shock_MEF_up, heat_shock_MEF_down, heat_shock_MEF_unchange)
# Int conversion
heat_shock_MEF$start <- as.integer(heat_shock_MEF$start)
heat_shock_MEF$end <- as.integer(heat_shock_MEF$end)
heat_shock_MEF$log10pvalue <- as.numeric(heat_shock_MEF$log10pvalue)


## Hypoxia HeLa Data 
hypoxia_HeLa_up <- readxl::read_excel(paste(data_folder, 
                                              'hypoxia_m6A_sites_HeLa.xlsx', 
                                              sep=''),
                                        sheet = 'Hypoxia UP m6A')
hypoxia_HeLa_down <- readxl::read_excel(paste(data_folder, 
                                                'hypoxia_m6A_sites_HeLa.xlsx', 
                                                sep=''),
                                          sheet = 'Hypoxia Down m6A')
hypoxia_HeLa_unchange <- readxl::read_excel(paste(data_folder, 
                                                    'hypoxia_m6A_sites_HeLa.xlsx', 
                                                    sep=''),
                                              sheet = 'Hypoxia un-change m6A')

hypoxia_HeLa <- rbind(hypoxia_HeLa_up, hypoxia_HeLa_down, hypoxia_HeLa_unchange)
# Int conversion
hypoxia_HeLa$start <- as.integer(hypoxia_HeLa$start)
hypoxia_HeLa$end <- as.integer(hypoxia_HeLa$end)
hypoxia_HeLa$log10pvalue <- as.numeric(hypoxia_HeLa$log10pvalue)
# Renaming direction since there is two values: Hypoxia and hypoxia 
hypoxia_HeLa$direction <- str_to_sentence(hypoxia_HeLa$direction)

# renaming
heat_shock_MEF <- heat_shock_MEF |> 
  rename('chr'='seqnames')
hypoxia_HeLa <- hypoxia_HeLa |> 
  rename('chr'='seqnames')
 
summary_data <- rbind( hypoxia_HeLa |> 
                         group_by(direction, label) |> 
                         count() |> ungroup() |> 
                         arrange(label),
                       heat_shock_MEF |>
                         group_by(direction, label) |> 
                         count() |> ungroup() |> 
                         arrange(label)) 

summary_data <- summary_data |>
  pivot_wider(names_from = label, values_from = n) |>
  rowwise(direction) |>
  mutate(total = sum(c(`Down-regulated m6A`,`Unchanged m6A`, `Up-regulated m6A`))) |>
  ungroup() |> 
  rename('label'='direction')

summary_data <- rbind(summary_data, list('HEK293T m6A Overlap Repl Sites', NA, NA, NA, dim(overlap_repl_sites)[1]))

print(summary_data)

rm(hypoxia_HeLa_up, hypoxia_HeLa_down, hypoxia_HeLa_unchange)
rm(heat_shock_MEF_up, heat_shock_MEF_down, heat_shock_MEF_unchange)
```

NOTE!!!: We can see there is a difference between the data provided and the GLORI paper. It seems that the Hypoxia and HS data are somehow interchanged.


# Counting m6A for each chromosomes 
```{r}
library(ggplot2)
## For plots or analysis
chr_order <- c('chr1','chr2','chr3','chr4','chr5','chr6','chr7',
                            'chr8','chr9','chr10','chr11','chr12','chr13','chr14',
                            'chr15','chr16','chr17','chr18','chr19','chr20','chr21',
                            'chr22','chrX','chrY','chrM') 
overlap_repl_count <- overlap_repl_sites |> 
  group_by(chr) |> 
  count() |> 
  mutate(data='repl_sites_HEK293T')
HeLa_count <- hypoxia_HeLa |> 
  group_by(chr) |> 
  count() |> 
  mutate(data='HeLa_hypox')
MEF_count <- heat_shock_MEF |> 
  group_by(chr) |> 
  count() |> 
  mutate(data='MEF_heat')

count_data <- rbind(HeLa_count, MEF_count, overlap_repl_count)
rm(HeLa_count, MEF_count, overlap_repl_count)


count_data <- count_data |> 
  pivot_wider(names_from = data, values_from = n) |> 
  mutate(chr = factor(chr, levels = chr_order)) |> 
  arrange(chr)

print(count_data)

```

Notice different count of m6A sites for each of the cell type of different conditions. In heat_shock, MEF doesn't have any m6A data at chr20, chr21, and chr22. It might be due to several reasons which were not covered on the paper. First, the sites might not have site either on the control or on the treatment environment. Second, it might be that those sites don't have any m6A sites.

However, when we see the intersection of replicates within a normal condition of HEK293T cells, those chromosomes have m6A site. Therefore, the former hypothesis might be the reason.

```{r}
# Create the density plot and the distribution of the value 
ggplot(overlap_repl_sites, aes(x = avg_m6a*100)) +
  geom_density(color = "brown", size = 1) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank()) +
  labs(x = "Methylation level (%)", y = "Density")

ggplot(overlap_repl_sites, aes(x = 'overlap sites', y = avg_m6a*100)) +  
  geom_violin(aes(fill="orange")) +     
  geom_boxplot(width=0.1) + 
  theme_bw() + 
  labs(y = "Methylation levle (%)")
```
We can see that the median of the m6A rate is ~40%. In addition, most of the sites are considered not to have methylation modification (<50%).


In addition, we see that neither HeLa hypoxia nor the MEF heat shock data contain gene annotation at it. Thus, we will annotate it using the longest transcript using the GLORI annotation tools found on its GitHub page. 


# Running the GLORI annotation pipeline using p.14 NCBI GTF file
```{bash}
# This script is used to create the annotation file for the human genome
# The annotation python scripts were downloaded from GLORI GitHub repository

# Download the human genome annotation file
#mkdir -p data/annot_ref/
#wget  https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/GCF_000001405.40-RS_2023_03/GCF_000001405.40_GRCh38.p14_assembly_report.txt 

#wget https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/GCF_000001405.40-RS_2023_03/GCF_000001405.40_GRCh38.p14_genomic.gtf.gz 

#conda activate thesis

# Patch the GTF file based on the assembly report
#python GLORI/get_anno/change_UCSCgtf.py -i data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf -j data/annot_ref/GCF_000001405.40_GRCh38.p14_assembly_report.txt -o data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf_change2Ens

# Get exon start and end
#python GLORI/get_anno/gtf2anno.py -i data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf_change2Ens -o data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf_change2Ens.tbl

# Filtering the unknown transcript
#awk '$3!~/_/&&$3!="na"' data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf_change2Ens.tbl | sed '/unknown_transcript/d'  > data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf_change2Ens.tbl2

# Get the longest transcript for each gene
#python GLORI/get_anno/selected_longest_transcrpts_fa.py -anno data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf_change2Ens.tbl2  --outname_prx data/annot_ref/GCF_000001405.39_GRCh38.p13_rna2
```



# Get gene annotation 
```{r}
# Filtered LONGEST TRANSCRIPT from h38.p14 GTF (GCF_000001405.40_GRCh38.p14_genomic.gtf.gz)
# See GLORI Github page for the pipeline of filtering longest transcript

longest_transcript <- read.table('data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gtf.translength.longest.trans',
                                 col.names=c('chr','start','end','length',
                                             'gene','strand','transcript'), skip=1)
hypoxia_HeLa_annot <- fuzzyjoin::genome_join(hypoxia_HeLa, longest_transcript, 
                                             by = c("chr", "start", "end"), 
                       mode = "left") 

hypoxia_HeLa_annot <- hypoxia_HeLa_annot |> 
  dplyr::select(!c('chr.y')) |> 
  rename('chr'='chr.x', 'start'='start.x', 'end'='end.x',
         'max_start'='start.y', 'max_end'='end.y')


heat_shock_MEF_annot <- fuzzyjoin::genome_join(heat_shock_MEF, longest_transcript, 
                                             by = c("chr", "start", "end"), 
                                             mode = "left") 

heat_shock_MEF_annot <- heat_shock_MEF_annot |> 
  dplyr::select(!c('chr.y')) |> 
  rename('chr'='chr.x', 'start'='start.x', 'end'='end.x',
         'max_start'='start.y', 'max_end'='end.y')

print(hypoxia_HeLa_annot)
print(heat_shock_MEF_annot)


```

We can see that there are several regions which have null value of the join result. It might be that the region is actually not containing any gene i.e. chr1  10806198 10806198 on heat shock.  
```{r}
# There might be several genes on a position. Check: 
print(heat_shock_MEF_annot |> group_by(chr, start, end) |> count() |> arrange(desc(n)))
print(heat_shock_MEF_annot |> filter(chr=='chr5', start==141410368, end==141410368))
print(hypoxia_HeLa_annot |> group_by(chr, start, end) |> count() |> arrange(desc(n)))
print(hypoxia_HeLa_annot |> filter(chr=='chr12', start==10996111, end==10996111))

# Count of non-annotated pos: 
print(heat_shock_MEF_annot |> filter(is.na(gene)))
print(hypoxia_HeLa_annot |> filter(is.na(gene)))
```


# %m6A Distribution of Hypoxia and HS 
```{r}

heat_shock_dist <- heat_shock_MEF |> 
  dplyr::select(c('meth_case','meth_control')) |> 
  pivot_longer(cols = starts_with('meth_'), names_to = 'label',
               values_to = 'm6A_rate') |> 
  mutate(data='HeLa_Hypoxia')

hypoxia_HeLa_dist <- hypoxia_HeLa |> 
  dplyr::select(c('meth_case','meth_control')) |> 
  pivot_longer(cols = starts_with('meth_'), names_to = 'label',
               values_to = 'm6A_rate') |> 
  mutate(data='MEF_Heat_Shock')

dist_data <- rbind(heat_shock_dist, hypoxia_HeLa_dist)

comps <- c('meth_control', 'meth_case')
dist_plot <- ggplot(dist_data,aes(x= label, y = m6A_rate)) + 
  geom_violin(aes(fill=label)) +     
  geom_boxplot(width=0.1) +  ggtitle(label = "Distribution Control vs Case") +
  facet_grid(cols = vars(data)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  theme_bw() +
  stat_compare_means(method = "wilcox.test") # Add two-sided test

dist_plot 

dist_data |> 
  group_by(data, label) |> 
  summarise(mean(m6A_rate), .groups = 'drop')
```

From the distribution of m6A rate per cases, we can see that Hypoxia in HeLa cell shifts the distribution of m6A_rate to be significantly higher compared to control (p-val <0.05). Moreover, there is also significant (p-val <0.05)  different in m6A_rate between the Heat Shock treated and control group. However, this overall different in distribution does not depict the stoichiometry since there might be increment of methylation on a particular regions and also decrement of methylation on other regions.


Next, we will show the volcano plot and analyse the significant genes
```{r}
heat_shock_delta_m6A <- heat_shock_MEF_annot |> 
  dplyr::select(c('meth_diff','log10pvalue', 'label', 'gene')) |> 
  mutate(data='HeLa_Hypoxia')

hypoxia_delta_m6A <- hypoxia_HeLa_annot |> 
  dplyr::select(c('meth_diff','log10pvalue', 'label', 'gene')) |> 
  mutate(data='MEF_Heat_Shock')

delta_m6A_data <- rbind(heat_shock_delta_m6A, hypoxia_delta_m6A)


ggplot(data = delta_m6A_data, mapping = aes(x = meth_diff, y = log10pvalue)) + 
  geom_point(alpha=0.25, mapping = aes(colour = label), size = 0.8)  +
  ylim(0, 20) +
  # From the paper: When the absolute change of m6A modification level between untreated and stressed was greater than 15% and the P value was less than 0.05, it was considered to be a dynamic m6A.
  geom_vline(xintercept=c(-15,15),lty=2,col="black") + # this for the 15% change both down and up regulated
  geom_hline(yintercept = -log10(0.05),lty=4,col="black") + # for the <0.05 p-value
  geom_text_repel(data = delta_m6A_data[delta_m6A_data$label!='Unchanged m6A' & !is.na(delta_m6A_data$gene),]
                  , mapping = aes(label = gene), max.overlaps = 35, size = 2,
                  point.size = NA) +
  scale_color_manual(values=c("blue", "gray", "red")) +
  #annotate("text", x = 65, y = 2, label = paste0("n = ", delta_m6A_data |>  count()), size = 3, fontface = "bold", color = "red") + # Label to horizontal cut-off line.
  facet_wrap(vars(data)) + 
  labs(x="delta %m6A",
       y=TeX(r"($-log_{10}p$)"),
       title="Stress Induced Response")  +
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="bottom", 
        legend.title = element_blank())
```

```{r}
# Print HSPA related genes 
print(heat_shock_MEF_annot |> 
        drop_na() |> 
        dplyr::select(c(chr,start,end,meth_diff,log10pvalue,label,gene)) |> 
        filter(grepl('(?i)hsp|HSPA|IRS3|VEGFA|EGLN', gene)) |> 
        arrange(chr, start, end)) 
  

# Print Hypoxia related genes 
print(hypoxia_HeLa_annot |> 
        drop_na() |> 
        dplyr::select(c(chr,start,end,meth_diff,log10pvalue,label,gene)) |> 
        filter(grepl('HSPA|IRS3|VEGFA|EGLN', gene)) |> 
        arrange(chr, start, end)) 
```



```{r}
# Print HSPA related genes 
print(heat_shock_MEF_annot |> 
        drop_na() |> 
        dplyr::select(c(chr,start,end,meth_diff,log10pvalue,label,gene)) |> 
        filter(grepl('HSPA', gene)) |> 
        dplyr::arrange(chr, start, end)) 
  

# Print Hypoxia related genes 
print(hypoxia_HeLa_annot |> 
        drop_na() |> 
        dplyr::select(c(chr,start,end,meth_diff,log10pvalue,label,gene)) |> 
        filter(grepl('IRS3|VEGFA|EGLN', gene)) |> 
        arrange(chr, start, end)) 
```




# Looking at the GRE data. HeLA and Hypox related RAW Data
Notes: GLORI used deep sequencing (~390,000,000 uniquely mapped reads, or ~140× coverage).
COLUMNS: 
CR --> Conversion rate for genes 
AGcov     --> reads coverage with A and G
Acov      --> read coverage with A
GeneCov   --> mean coverage for the whole genome 
Ratio     --> A rate for the sites (methylation level)
Pvalue    --> test for A rate based on the background (binomial distribution or Poisson distribution, to each of them on the basis of the gene-specific A-to-G conversion rate)
P_adjust  --> FDR adjusted Pvalue (BH method)

The number of m6A reads for each detected m6A site was calculated as A rate × cov- erage. The methylation level for one gene was then calculated as: the sum of m6A reads of all m6A sites divided by the sum of coverage of all sites in the longest isoform.

# Prepare data
```{r}
data_path <- '/Users/faqih/Documents/UCPH/Thesis/code/data/'

# A rate serving as the m6A level
GRE_folder = '/Users/faqih/Documents/UCPH/Thesis/papers/GLORI/GLORI_data/GSE210563_RAW/'
hela_1 <- read.csv(paste0(GRE_folder,'GSM6432595_Hela-1_35bp_m2.totalm6A.FDR.csv'), sep = "\t")
hela_2 <- read.csv(paste0(GRE_folder,'GSM6432596_Hela-2_35bp_m2.totalm6A.FDR.csv'), sep = "\t")
mef_1 <- read.csv(paste0(GRE_folder,'GSM6432601_MEF-1_35bp_m2.totalm6A.FDR.csv'), sep = "\t")
mef_2 <- read.csv(paste0(GRE_folder,'GSM6432602_MEF-2_35bp_m2.totalm6A.FDR.csv'), sep = "\t")

hypox_hela_1 <- read.csv(paste0(GRE_folder,'GSM6432598_hypoxia-1_35bp_m2.totalm6A.FDR.csv'), sep = '\t')
hypox_hela_2 <- read.csv(paste0(GRE_folder,'GSM6432599_hypoxia-2_35bp_m2.totalm6A.FDR.csv'), sep = '\t')
hs_mef_1 <- read.csv(paste0(GRE_folder,'GSM6432604_HS-1_35bp_m2.totalm6A.FDR.csv.gz'), sep = '\t')
hs_mef_2 <- read.csv(paste0(GRE_folder,'GSM6432605_HS-2_35bp_m2.totalm6A.FDR.csv.gz'), sep = '\t')

#NOTES:
#   hypox_hela_2 |>  inner_join(heat_shock_MEF, by = dplyr::join_by(Chr==chr, Sites==start)) RETURNING SMALL NUMBER OF SITES (4)
#   hypox_hela_2 |>  inner_join(hypoxia_HeLa, by = dplyr::join_by(Chr==chr, Sites==start)) RETURNING >80K sites. THUS HYPOXIA_HELA IS INDEED HYPOXIA? 
# This is conversely the sdame for the MEF-HS.
# Conclusion: the supplementary tables are indeed true. The publisehd figures' labels were False? 

# Write cov data. Required for methylSig beta binom diffd 
write.table(hela_1 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/hela_1.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(hela_2 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/hela_2.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(hypox_hela_1 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/hypox_hela_1.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(hypox_hela_2 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/hypox_hela_2.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)


write.table(mef_1 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/mef_1.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(mef_2 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/mef_2.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(hs_mef_1 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/hs_mef_1.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(hs_mef_2 |> 
              mutate(Ratio = as.numeric(Ratio) * 100, start=Sites, end=Sites) |> 
              select(Chr, start, end, Ratio, Acov, AGcov),  # Ensure Ratio is numeric
            file = paste0(data_path, "GLORI_data/hs_mef_2.cov"), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)

# Note: ELSE for gene_name and transcript represents unknown gene and/or transcript.
```

```{r}
print(dim(hela_1))
print(dim(hela_2))
print(dim(hypox_hela_1))
print(dim(hypox_hela_2))
```

# Hela-Hypoxia
When the absolute change of m6A modification level between untreated and stressed was greater than 15% and the P value was less than 0.05, it was considered to be a dynamic m6A.
```{r}
#hypoxia_HeLa
#chr    start    end meth_case meth_control meth_diff direction    pvalue      fdr log_lik_ratio  log10pvalue label
#chr1  907102 907102      66.1         38.2      27.9 Hypoxia   0.000999  0.00632     10.8    3.00   Up-regula…
hela_1  |> filter(Chr == "chr1", Sites == 185038) # Ratio 0.42553 0.411662
hela_2 |> filter(Chr == "chr1", Sites == 185038) # Ratio 0.33333 0.3214235
hypox_hela_1 |> filter(Chr == "chr1", Sites == 185038) # Ratio 0.68182 NormeRatio 0.6509131
hypox_hela_2 |> filter(Chr == "chr1", Sites == 185038) # Ratio 0.64706 NormeRatio 0.6290264	

# Meean hypox hela 0.664
# mean hela 0.42553 
```

```{r}
# List of file paths
library(methylSig)
library(bsseq)

file_paths <- c(
                paste0(data_path, "GLORI_data/hela_1.cov"), 
                paste0(data_path, "GLORI_data/hela_2.cov"), 
                paste0(data_path, "GLORI_data/hypox_hela_1.cov"), 
                paste0(data_path, "GLORI_data/hypox_hela_2.cov"))  # Corrected to include hypox_hela_2.cov

# Corrected colData with a row for each file
colData <- DataFrame(condition = c(
                                   "control", 
                                   "control", 
                                   "hypox", 
                                   "hypox"),
                     replicate = c(
                                   1, 
                                   2, 
                                   1, 
                                   2))
row.names(colData) <- c(
                        "hela_1",
                        "hela_2", 
                        "hypox_hela_1",
                        "hypox_hela_2")
# Ensure colData has the same number of rows as there are files
if (length(file_paths) != nrow(colData)) {
    stop("The number of files does not match the number of rows in colData.")
}

# Read the files with bsseq::read.bismark
bsseq_data <- bsseq::read.bismark(file_paths, 
                                   BPPARAM = BiocParallel::MulticoreParam(workers = 3, 
                                                                          progressbar = FALSE),
                                  verbose=TRUE,
                                  )
pData(bsseq_data) <- colData
pData(bsseq_data)
# Require at least two samples from control and two samples from hypox
bsseq_data_filtered = filter_loci_by_group_coverage(
    bs = bsseq_data,
    group_column = 'condition',
    c('control' = 2, 'hypox' = 2))
# Methylation matrix
#head(bsseq::getCoverage(combined_BSseq_filtered, type = 'M'))
#GenomicRanges::granges(combined_BSseq_filtered)

diff_gr_hypox = diff_methylsig(
    bs = bsseq_data_filtered,
    disp_groups = c('case' = TRUE, 'control' = TRUE),
    group_column = 'condition',
    comparison_groups = c('case' = 'hypox', 'control' = 'control'),
    n_cores=3)

diff_gr_hypox
```
```{r}
diff_df_hypox <- as.tibble(as.data.frame(diff_gr_hypox))
diff_df_hypox |> inner_join(hypoxia_HeLa, by = dplyr::join_by(seqnames==chr, start==start, end==end))
```




# MEF-Heatshock 
```{r}
file_paths <- c(
                paste0(data_path, "GLORI_data/mef_1.cov"), 
                paste0(data_path, "GLORI_data/mef_2.cov"), 
                paste0(data_path, "GLORI_data/hs_mef_1.cov"), 
                paste0(data_path, "GLORI_data/hs_mef_2.cov")) 

colData <- DataFrame(condition = c(
                                   "untreated", 
                                   "untreated", 
                                   "hs", 
                                   "hs"),
                     replicate = c(
                                   1, 
                                   2, 
                                   1, 
                                   2))
row.names(colData) <- c(
                        "mef_1",
                        "mef_2", 
                        "hs_mef_1",
                        "hs_mef_2")
# Ensure colData has the same number of rows as there are files
if (length(file_paths) != nrow(colData)) {
    stop("The number of files does not match the number of rows in colData.")
}

# Read the files with bsseq::read.bismark
bsseq_data <- lapply(file_paths, bsseq::read.bismark)


# Assuming bsseq_data is your list of BSseq objects
combined_BSseq <- do.call(combine, bsseq_data)
pData(combined_BSseq) <- colData


combined_BSseq_filtered = filter_loci_by_group_coverage(
    bs = combined_BSseq,
    group_column = 'condition',
    c('untreated' = 2, 'hs' = 2))

diff_gr_hela = diff_methylsig(
    bs =
      #combined_BSseq,
      combined_BSseq_filtered,
    disp_groups = c('case' = TRUE, 'control' = TRUE),
    group_column = 'condition',
    comparison_groups = c('case' = 'hs', 'control' = 'untreated'),
    n_cores=3)

diff_gr_hela
```

Results are smaller? 




# Creating the geneannotation from ENSEMBL
```{r}
# GFF3 data from ENSEMBL FTP 
human_gff_path_ensembl <- 'data/annot_ref/Homo_sapiens.GRCh38.111.gff3'
# Creating ensembldB (MariaDb)
GFF_dB <- ensembldb::ensDbFromGff(human_gff_path_ensembl)
# Converting into EnsDb object
human_TxDb_ensembl <- ensembldb::EnsDb(GFF_dB)
# Get the transcript length from TxDb
human_Tx_length_ensembl <- transcriptLengths(human_TxDb_ensembl,
                  with.cds_len=TRUE, with.utr5_len=TRUE, with.utr3_len=TRUE)

# NOTES:
# 'tx_len' in a lot of cases will not be the same as the 'width' column.
#   The reason behind it is that tx_len not consider intron length.
#   width of a transcript includes the Introns! while tx_len includes cds, utr5 and utr 3 only
# cds_len, utr5_len, utr3_len, hvae values of zero if bio
# tx_cds_seq_start/end has value of NA ()
# Remember that Transcription start upstream of TATA Box (~30%). It also contain several things such as 
# BRE(upstream of TATA) which acts as TFIIB recognition element, INR (initiator element)
# or other promoters such as MTE and DTE which upstream the TSS.
temp_info <- transcripts(human_TxDb_ensembl, 
                         filter=TxIdFilter(rownames(human_Tx_length_ensembl)), 
                         columns=c("tx_name", "gene_id", "gene_name", 
                                   "gene_biotype", 'tx_seq_start', 
                                   'tx_cds_seq_start', 'tx_cds_seq_end', 
                                   'tx_seq_end'))

human_Tx_ensembl_data <- S4Vectors::merge(human_Tx_length_ensembl, temp_info, 
                                    by.x = 'tx_id', by.y = 'tx_name', all.x=TRUE)

# Filtering the longest transcript (width) only for each gene
# it still contain numbers of non mRNA gene: i.e. rRNA, tRNA, Mt_RNA, etc.
human_Tx_ensembl_data <- human_Tx_ensembl_data |> 
  dplyr::group_by(gene_id.x) |> 
  dplyr::filter(width == max(width)) |> 
  # Since several transcripts could have the same start and end.
  # need to filter by tx_len (CDS + UTRs)
  dplyr::filter(tx_len == max(tx_len)) |> 
  dplyr::ungroup() |> 
  dplyr::select(!c(tx_name, gene_id.y, tx_id.y)) |> 
  dplyr::rename('gene_id'='gene_id.x', 'chr'='seqnames') |> 
  dplyr::relocate(c('gene_name','gene_biotype'), .after = c('gene_id')) |> 
  dplyr::relocate(c('tx_id','chr','start','end','width','strand'), 
                  .after = c('gene_biotype'))

rm(temp_info, human_Tx_length_ensembl)
```

```{r}
# Save the guitarTxdb to be loaded later
save(human_TxDb, file = 'human_GuitarTxdb')
```


```{r}
human_TxDb <-  makeTxDbFromGFF(file=human_gff_path_ensembl, format = 'gff3',
                               organism = 'Homo sapiens')
human_GuitarTxdb <- Guitar::makeGuitarTxdb(human_TxDb)

# Save the guitarTxdb to be loaded later
save(human_TxDb, file = 'human_GuitarTxdb')
```

```{r}
## NED TO CHANGE INTO ENSEMBL MODE # 1,2, MT, Y, X
temp_m6A_bed <- overlap_repl_sites |> 
  dplyr::mutate(stop = sites) |> 
  dplyr::rowwise() |> 
  # Change into ENSEMBL format
  dplyr::mutate(name = paste0('m6A_',chr,'_',sites), 
                CHR=ifelse(strsplit(as.character(chr), "chr")[[1]][2] == "M",
                       "MT", 
                       strsplit(as.character(chr), "chr")[[1]][2]),
                avg_m6a = avg_m6a*100) |> 
  dplyr::ungroup() |> 
  #dplyr::arrange(CHR,sites,stop) |> 
  dplyr::rename(start=sites) |> 
  dplyr::select(CHR, start, stop, name, avg_m6a, strand) 

# Writing into bed file 
m6A_overlap_bed <- GenomicRanges::makeGRangesFromDataFrame(temp_m6A_bed)
m6A_overlap_bed <- rtracklayer::export(m6A_overlap_bed, format="bed", ignore.strand = FALSE)
write.table(m6A_overlap_bed, "m6A_overlap.bed", sep="\t", col.names=FALSE,
            row.names = FALSE, quote = FALSE)

# Overlapping with 
```


```{r}
  stSampleNum = 10
mapFilterTranscript = TRUE
  sitesGroup <- .getStGroup(stBedFiles = "m6A_overlap.bed", stGRangeLists = NULL, 
                            stGroupName = NULL)
  GroupNames <- names(sitesGroup)
  
  
  sitesGroupNum <- length(sitesGroup)
  sitesPointsNormlize <- list()
  sitesPointsRelative <- list()
  pointWeight <- list()
  for (i in seq_len(sitesGroupNum)) {
    GroupName = GroupNames[[i]]
    print(paste("sample", stSampleNum, "points for" , GroupName, sep = " "))
    sitesPoints <- samplePoints(sitesGroup[i], 
                                stSampleNum = stSampleNum,
                                stAmblguity = stAmblguity,
                                pltTxType =  pltTxType, 
                                stSampleModle = stSampleModle, 
                                mapFilterTranscript = mapFilterTranscript,
                                guitarTxdb)
    
    for (txType in pltTxType) {
      sitesPointsNormlize[[txType]][[GroupName]] <- normalize(sitesPoints, guitarTxdb, txType,overlapIndex,siteLengthIndex)
      sitesPointsRelative[[txType]][[GroupName]] <- sitesPointsNormlize[[txType]][[GroupName]][[1]]
      pointWeight[[txType]][[GroupName]] <- sitesPointsNormlize[[txType]][[GroupName]][[2]]
    }
  }
```

```{r}
gcl <- list(human_GuitarTxdb$tx$tx)
GuitarPlot(txTxdb = human_TxDb,
stGRangeLists = gcl,
stSampleNum = 200,
enableCI = TRUE,
pltTxType = c("tx"),
txPrimaryOnly = FALSE
)
```



```{r}
# if Error in reduce(txWithFlank) : argument ".f" is missing, with no default --> unload purrr package. conflict namespace
GuitarPlot(txTxdb = human_TxDb,
          #mapFilterTranscript=FALSE,
           stBedFiles = 'm6A_overlap.bed', enableCI=FALSE)
```

```{r}
coba <- function(sitesGRanges, guitarTxdb, txType,overlapIndex,siteLengthIndex)
{ 
  sitesInformation <- data.frame()
  
  sitesPointsPositionNormalize <- list()
  
  sitesPointsPosTx<-list()
  
  sitesPointsPosTx<-end(sitesGRanges[[txType]])
  
  names(sitesPointsPosTx)<-seqnames(sitesGRanges[[txType]])
  #step 1
  startPointMat <- guitarTxdb[[txType]]$startPoint[names(sitesPointsPosTx), ]
  startPointDiffer <- startPointMat - sitesPointsPosTx
  max_which <- function(x)
  {
    max(which(x<0))
  }
  component_which <- function(x, Componet_pct, sitesPointsComponet)
  {
    sitesPointsComponet_pct <- Componet_pct[x,][sitesPointsComponet[[x]]]
  }
  sitesPointsComponet <- apply(startPointDiffer, 1, max_which)
  # step 2
  sitesPointsPositionComponet<-startPointDiffer[cbind(seq_along(sitesPointsComponet), sitesPointsComponet)] * -1
  # step 3
  sitesPointsComponetMat<-startPointMat[cbind(seq_along(sitesPointsComponet), sitesPointsComponet)]
  # step 4
  sitesPointsComponetWidthAvg <- guitarTxdb[[txType]]$componentWidthAverage_pct[sitesPointsComponet]
  # step 5
  sitesPointsComponetStart_pct <- guitarTxdb[[txType]]$componentStartAverage_pct[sitesPointsComponet]
  # step 6
  componentWidthMat <- guitarTxdb[[txType]]$componentWidth[names(sitesPointsPosTx), ]
  sitesPointsComponetWidth <- componentWidthMat[cbind(seq_along(sitesPointsComponet), sitesPointsComponet)]
  # step 7
  sitesPointsPositionNormalize <- sitesPointsPositionComponet / sitesPointsComponetWidth * sitesPointsComponetWidthAvg + sitesPointsComponetStart_pct
  names(sitesPointsPositionNormalize) <- sitesGRanges[[txType]]$xHits
  #step 8 
  sitesComponet_pct <- guitarTxdb[[txType]]$componentWidthPtc[names(sitesPointsComponet),]
  sitesPointsComponet_pct <- unlist(lapply( 1:length(sitesPointsComponet), component_which, sitesComponet_pct, sitesPointsComponet))
  sitesPointsWeight <-  sitesPointsComponetWidthAvg / (sitesGRanges[[txType]]$pointsOverlapTx^overlapIndex )/ sitesPointsComponet_pct * (sitesGRanges[[txType]]$sitesLength ^ siteLengthIndex)
  names(sitesPointsWeight) <- sitesGRanges[[txType]]$xHits 
  return(list(sitesPointsPositionNormalize,sitesPointsWeight))
}


coba(m6A_overlap_bed,human_GuitarTxdb, "mRNA", 1, 1)
```



# FROM NCBI 
```{r}
# Get GFF 
# Incorrect chr info for NCBI gff? https://www.biostars.org/p/9552902/
human_gff_ncbi_path <- 'data/annot_ref/GCF_000001405.40_GRCh38.p14_genomic.gff'
human_gff_ncbi_report_path <- 'data/annot_ref/GCF_000001405.40_GRCh38.p14_assembly_report.txt'

# Load transcripts DB
human_TxDb_ncbi <- makeTxDbFromGFF(file=human_gff_ncbi_path, format = 'gff3',
                                   organism = 'Homo sapiens')

# NOTE: GFF from NCBI doesn't contain straight forward chromosome name
# Thus we need to use the report assembly and the seqnames (RefSeq-Accn) 
# into UCSC-style-name later
human_gff_ncbi_report <- read.csv(human_gff_ncbi_report_path, comment = '#', 
                                  sep='\t', header = FALSE, 
                                  col.names = c("Sequence-Name","Sequence-Role",
                                                "Assigned-Molecule",
                                                "Assigned-Molecule-Location/Type",
                                                "GenBank-Accn","Relationship",
                                                "RefSeq-Accn","Assembly-Unit",
                                                "Sequence-Length","UCSC-style-name")
                                  )

# Rename the seqnames (chr) into UCSC style
temp_seq_lvl <- data.frame(seqlevels = seqlevels((human_TxDb_ncbi)))
temp_seq_lvl <- temp_seq_lvl |> 
  dplyr::left_join(human_gff_ncbi_report |> dplyr::select(RefSeq.Accn, UCSC.style.name),
                   by = dplyr::join_by(seqlevels==RefSeq.Accn)) |> 
  dplyr::pull(UCSC.style.name)
# Renaming
seqlevels(human_TxDb_ncbi) <- temp_seq_lvl


# load transcripts GR
human_Tx_ncbi <- transcripts(human_TxDb_ncbi, columns = c("tx_id", "tx_name", "TXTYPE"))
# get length of transcripts
human_Tx_length_ncbi <- transcriptLengths(human_TxDb_ncbi,
                  with.cds_len=TRUE, with.utr5_len=TRUE, with.utr3_len=TRUE)

# Calculating CDS start/end. Reference: https://support.bioconductor.org/p/9154904/
cds_start <- start(human_Tx_ncbi) +
             ifelse(strand(human_Tx_ncbi) == "+", human_Tx_length_ncbi$utr5_len, 
                    human_Tx_length_ncbi$utr3_len)
cds_end   <- end(human_Tx_ncbi)   -
             ifelse(strand(human_Tx_ncbi) == "+", human_Tx_length_ncbi$utr3_len, 
                    human_Tx_length_ncbi$utr5_len)

no_cds_idx <- which(human_Tx_length_ncbi$cds_len == 0L)
cds_start[no_cds_idx] <- NA
cds_end[no_cds_idx]   <- NA

mcols(human_Tx_ncbi)$cds_start <- cds_start
mcols(human_Tx_ncbi)$cds_end   <- cds_end

human_Tx_ncbi_data <- S4Vectors::merge(human_Tx_length_ncbi, human_Tx_ncbi, 
                                       by.x = 'tx_id', by.y = 'tx_id', all.x=TRUE)


# Filtering the longest transcript (width) only for each gene
# it still contain numbers of non mRNA gene: i.e. rRNA, tRNA, Mt_RNA, etc.
human_Tx_ncbi_data <- human_Tx_ncbi_data |> 
  dplyr::group_by(gene_id) |> 
  dplyr::filter(width == max(width)) |> 
  # Since several transcripts could have the same start and end.
  # need to filter by tx_len (CDS + UTRs)
  dplyr::filter(tx_len == max(tx_len)) |> 
  dplyr::ungroup() |> 
  dplyr::select(!c(tx_name.y)) |>
  dplyr::rename('gene_biotype'='TXTYPE',
                'tx_name'='tx_name.x', 'chr'='seqnames') |>
  dplyr::relocate(c('gene_biotype'), .after = c('gene_id')) |>
  dplyr::relocate(c('tx_name','chr','start','end','width','strand'),
                  .after = c('gene_biotype'))

AnnotationDbi::select(human_TxDb_ncbi, keys=human_Tx_ncbi_data$tx_name, 
                      columns='GENEID', keytype = 'TXNAME')

longest_tx <- list(tx_id = human_Tx_ncbi_data |> dplyr::pull(tx_id))
longest_mRNA_tx <- list(tx_id = human_Tx_ncbi_data |> 
                          dplyr::filter(gene_biotype=='mRNA')
                          dplyr::pull(tx_id)


exons(human_TxDb_ncbi,filter=longest_tx, columns=c('exon_id', 'tx_name'))



  

# rm(cds_start, cds_end, no_cds_idx, human_Tx_length_ncbi, human_Tx_ncbi)
```
# Join with overlap
```{r}
txfiveutrMinLength = 100
txcdsMinLength = 100
txthreeutrMinLength = 100
txlongNcrnaMinLength = 100
txlncrnaOverlapmrna = FALSE
txpromoterLength = 1000
txtailLength = 1000
txAmblguity = 5
pltTxType =  c("tx","mrna","ncrna")
component <- list()
component$txTypes <- list()

nameFilterTx <- longest_tx$tx_id

tx <- exonsBy(human_TxDb_ncbi, by = "tx")
tx <- tx[longest_tx$tx_id]
txRange <- GenomicRanges::ranges(tx)

##################################################
  # generate components for all transcripts
  ##################################################
  # extract promoter and tail regions
  promoter <- flank(txRange, txpromoterLength, start=TRUE)
  tail <- flank(txRange, txtailLength, start=FALSE)
  
  # contruct transcripts with promoter and tail
  txGRange <- unlist(tx)
  mcols(txGRange) <- NULL    #promoter and tail don't have metadata
  txWithFlank_gr <- c(as(promoter, "GRanges"), txGRange,as(tail, "GRanges"))
  txWithFlank <- split(txWithFlank_gr, names(txWithFlank_gr))
  txWithFlank <- reduce(txWithFlank)
  
  for(txTypes in pltTxType )
  {
    if(txTypes == "tx" )
    {
      
      ##################################################
      # generate components for all tx
      ##################################################
      print("generate components for all tx")
      {
        component$txTypes <- c(component$txTypes, "tx")
        #component$tx$componentTypes <- ()
        component[[txTypes]]$componentTypes <- c("promoter", "rna", "tail")
        
        component[[txTypes]]$names <- nameFilterTx
        component[[txTypes]]$txWithFlank <- txWithFlank
        component[[txTypes]]$txWithFlank_len <- sum(width(component$tx$txWithFlank))
        component[[txTypes]]$tx <- tx
        component[[txTypes]]$promoter <- promoter
        component[[txTypes]]$rna <- tx
        component[[txTypes]]$tail <- tail
        
        component[[txTypes]]$txRange <- txRange
      }
    }
    ##################################################
    # generate components for mRNA
    ##################################################
    if(txTypes == "mrna" )
    {
      print("generate components for mRNA")
      if (length(longest_tx) > 0)
      { 
        component$txTypes <- c(component$txTypes, "mrna")
        component[[txTypes]]$componentTypes <- c("promoter", "utr5", "cds", "utr3", "tail")
        
        component[[txTypes]]$names <- longest_tx
        component[[txTypes]]$txWithFlank <- txWithFlank[longest_tx]
        component[[txTypes]]$txWithFlank_len <- sum(width(component$mrna$txWithFlank))
        component[[txTypes]]$tx <- tx[longest_tx]
        component[[txTypes]]$promoter <- promoter[longest_tx]
        component[[txTypes]]$utr5 <- utr5[longest_tx]
        component[[txTypes]]$cds <- cds[longest_tx]
        component[[txTypes]]$utr3 <- utr3[longest_tx]
        component[[txTypes]]$tail <- tail[longest_tx]
      }
    }
    ##################################################
    # generate components for lncRNA
    ##################################################
    if(txTypes == "ncrna" )
    {
      print("generate components for lncRNA")
      if (length(nameFilterncRNA) > 0)
      {
        component$txTypes <- c(component$txTypes, "ncrna")
        component[[txTypes]]$componentTypes <- c("promoter", "ncrna", "tail")
        
        component[[txTypes]]$names <- nameFilterncRNA
        component[[txTypes]]$txWithFlank <- txWithFlank[nameFilterncRNA]
        component[[txTypes]]$txWithFlank_len <- sum(width(component$ncrna$txWithFlank))
        component[[txTypes]]$tx <- tx[nameFilterncRNA]
        component[[txTypes]]$promoter <- promoter[nameFilterncRNA]
        component[[txTypes]]$ncrna <- tx[nameFilterncRNA]
        component[[txTypes]]$tail <- tail[nameFilterncRNA]
      }
    }
  }
  
  ##################################################
  # generate chiped transcriptome
  ##################################################
  print("generate chiped transcriptome")
  rslt <- .generateChipedTranscriptome(component)
  #print(paste(names(rslt), collapse = ", "))
  component$tx$txComponentGRange <- rslt$txComponentGRange
  component$tx$txChipedGRange <- rslt$txChipedGRange
  
```




# Sanity check from RAW GLORI with NCBI GFF3
```{r}
# Checking the inetrsection from RAW GLORI with NCBI p.14
join_raw_GLORI_ncbi <- hypox_hela_1 |> 
  dplyr::left_join(human_Tx_ncbi_data |> 
                     dplyr::select(gene, gene_biotype) |> 
                     dplyr::distinct(gene, gene_biotype), 
                   by = dplyr::join_by(Gene == gene), keep = TRUE)

na_type <- join_raw_GLORI_ncbi |> 
  dplyr::filter(is.na(gene), Gene != 'ELSE')
na_type_gene <- join_raw_GLORI_ncbi |> 
  dplyr::filter(is.na(gene), Gene != 'ELSE') |> 
  dplyr::distinct(Gene)

print(dim(na_type_gene))


# Check in GTF file 
transcripts(human_TxDb_ncbi, columns = c("tx_id", "tx_name", "TXTYPE"),  
            filter=list(gene_id=na_type_gene))
# Nothing. Don't know from where GLORI got additional 930 genes
# But mostly it is pseudogene. I think it comes from different patch version
# p.13 vs p.14
```
Question: why there' sstill missing Genes? when check 



# Metagene analysis of NCBI data 
```{r}
human_Tx_ncbi_premRNA_data <- human_Tx_ncbi_data |>  
  dplyr::filter(!is.na(cds_start)) |> 
  dplyr::rowwise() |> 
  dplyr::mutate(
    utr5_start = start,
    utr5_end =  ifelse(strand == "+", start + utr5_len - 1,
                       start - utr5_len + 1),
    utr3_start = ifelse(strand == "+", cds_end + 1,
                        cds_end - 1),
    utr3_end = end
  ) |> 
  dplyr::ungroup() |> 
  dplyr::relocate(c('utr5_start','utr5_end'), .before = c('cds_start'))

# Now, we want to create relative position of the sites within each gene
# We want to have postiion 0-1 as utr3, 1-2 as cds, and 2-3 as utr5, thus
# we can plot the relative position.

# But first we need to annot

# A tibble: 22,966 × 19
   gene    gene_biotype tx_id          chr     start     end width strand nexon tx_len cds_len utr5_len utr3_len utr5_start utr5_end cds_start cds_end utr3_start utr3_end
   <chr>   <chr>        <chr>          <chr>   <int>   <int> <int> <fct>  <int>  <int>   <int>    <int>    <int>      <int>    <dbl>     <int>   <int>      <dbl>    <int>
 1 OR4F5   mRNA         NM_001005484.2 chr1    65419   71585  6167 +          3   2618     981       60     1577      65419    65478     65479   70008      70009    71585
   
# say we are on 65468 Thus, from start we have: 65468 - utr5_start = 49 
# (49/width)*3 = 0.02383655  

```



# Load NCBI mapped sequence
```{r}
# Docs: https://www.bioconductor.org/packages/2.13/bioc/vignettes/Rsamtools/inst/doc/Rsamtools-Overview.pdf

# Known Refseq accessions prefixed with NM_ and NR_
human_known_seq_ncbi_path <- 'data/seq_ref/GCF_000001405.40_GRCh38.p14_knownrefseq_alns.bam'
# Model Refseq accessions prefixed with XM_ and XR_
# TODO human_model_seq_ncbi_path

# sam file: qname, flag, rname, pos, ...
human_known_seq_ncbi <- scanBam(human_known_seq_ncbi_path)

# Showing seqname example 
# human_known_seq_ncbi[[1]]$rname

# Showing tx name 
human_known_seq_ncbi[[1]]$qname

# TODO: Joining sequence data into the annotation data for training and testeing purpose
```












 
# ENSMBL BiomaRt scratch
```{r echo=FALSE, Univariate_Plots}
write.table(listAttributes(mart), file='list_attr.txt')
coba <- getBM(attributes = c('ensembl_transcript_id', '5_utr_start', '5_utr_end',
                                        '3_utr_start', '3_utr_end', 'genomic_coding_start', 
                                        'genomic_coding_start'),
                         filters = 'ensembl_gene_id',
                         values = 'ENSG00000272438',
                         mart = mart)
# TTN example
getBM(
     attributes = c('hgnc_symbol', 'gene_biotype', 'ensembl_gene_id', 'ensembl_transcript_id', 'ensembl_transcript_id_version'),
     filters = c('chromosome_name', 'start', 'end'),
     values = list(2, 178807422, 178807422),
     mart
 )

# get start stop for each transcript id 
# TTN example
getBM(
     attributes = c('hgnc_symbol','5_utr_start', '5_utr_end'),
     filters = 'ensembl_gene_id',
     values = 'NM_001267550.2',
     mart
 )


attributes <- c('ensembl_transcript_id', 
                '5_utr_start', '5_utr_end', 
                '3_utr_start', '3_utr_end')

# For the sake of example, let's assume `ensembl_transcript_ids` contains the mapped IDs
ids <- c('NM_001101', 'NM_001256799', 'NM_000594')

getBM(attributes = attributes,
                 filters = 'refseq_mrna',
                 values = ids,
                 mart = mart)
```
